### PR TITLE
🔥 feat(utils): add IsZeroValue

### PR DESCRIPTION
**Pull request by sixcolors**
## Description

Add utils function to support [change](https://github.com/gofiber/fiber/pull/2751#issuecomment-1852086176) to #2751

Adds IsZeroValue function to utils pkg. Which reports whether x is the zero value for its type. For basic types and error, it uses a fast path without reflection. For other types, it uses reflection which is slower.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] If new dependencies exist, I have checked that they are really necessary and agreed with the maintainers/community (we want to have as few dependencies as possible)
- [x] I tried to make my code as fast as possible with as few allocations as possible
- [x] For new code I have written benchmarks so that they can be analyzed and improved

### CODE REVIEW

**Comment by sixcolors** [comment]
basic types based on https://go.dev/tour/basics/11

File: utils/zero.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		return len(v) == 0
+	case []complex64:
+		return len(v) == 0
+	case []complex128:
+		return len(v) == 0
+	case []string:
+		return len(v) == 0
+	default:
+		// Slow path using reflection
+		return reflect.DeepEqual(x, reflect.Zero(reflect.TypeOf(x)).Interface())
```
Comment (efectn) [inline utils/zero.go:86]:
```go
type test string

func main() {
	x := []test{}
	fmt.Print(IsZeroValue(x))
}
```
will print false due to DeepEqual is pretty strict. Maybe this part should be adjusted?

**Comment by sixcolors** [comment]
If we were keen to add maps to the fast path we could make a generator, but that would add a lot of code, so I did not do it.

Generator code for reference:

```go
package main

import (
    "fmt"
    "os"
    "text/template"
)

type Type struct {
    Key   string
    Value string
}

func main() {
    types := []string{
        "bool", "int", "int8", "int16", "int32", "int64", "uint", "uint8", "uint16", "uint32", "uint64", "uintptr", "float32", "float64", "complex64", "complex128", "string", "error",
    }

    tmpl := template.Must(template.New("").Parse(`
    case map[{{.Key}}]{{.Value}}:
        return len(v.(map[{{.Key}}]{{.Value}})) == 0
    `))

    for _, keyType := range types {
        for _, valueType := range types {
            t := Type{Key: keyType, Value: valueType}
            err := tmpl.Execute(os.Stdout, t)
            if err != nil {
                fmt.Println("executing template:", err)
            }
        }
    }
}
```

File: utils/zero.go (hunk 1 above)
Comment (efectn) [reply utils/zero.go:86]: maybe this check is OK?
```go
val := reflect.ValueOf(x)
if val.Kind() == reflect.Slice && val.Len() == 0 {
	return true
}
```

Comment (efectn) [reply utils/zero.go:86]: > maybe this check is OK above this part?
> 
> ```go
> val := reflect.ValueOf(x)
> if val.Kind() == reflect.Slice && val.Len() == 0 {
> 	return true
> }
> ```

Comment (sixcolors) [reply utils/zero.go:86]: AFAIK in Go, a zero value for a slice is nil. An empty slice (like `x := []test{}` in your example) is not considered a zero value; it's just an empty slice.

Comment (sixcolors) [reply utils/zero.go:86]: I already considered empty slices zero in the fast path. So that would have to be the same... I'll fix.

Comment (efectn) [reply utils/zero.go:86]: > AFAIK in Go, a zero value for a slice is nil. An empty slice (like `x := []test{}` in your example) is not considered a zero value; it's just an empty slice.

Then we need to adjust switch-case part, right?

Comment (sixcolors) [reply utils/zero.go:86]: Which behaviour would you consider correct?

Comment (sixcolors) [reply utils/zero.go]: [cdb2885](https://github.com/gofiber/fiber/pull/2760/commits/cdb288542b077ecf5bb764d1d0033ba659c7a1dc) handles edge cases such as `x:= []test{}` and improves test/bench.

If we are going to consider those cases ZeroValues, otherwise we can revert to using DeepEqual.

Comment (efectn) [reply utils/zero.go]: > > > AFAIK in Go, a zero value for a slice is nil. An empty slice (like `x := []test{}` in your example) is not considered a zero value; it's just an empty slice.
> > 
> > 
> > Then we need to adjust switch-case part, right?
> 
> Correct.
> 
> [cdb2885](https://github.com/gofiber/fiber/pull/2760/commits/cdb288542b077ecf5bb764d1d0033ba659c7a1dc) handles edge cases such as `x:= []test{}` and improves test/bench.
> 
> If we are going to consider those cases ZeroValues we can keep it, otherwise we can revert to using DeepEqual and remove the zero length slice cases.

That's OK to me but not sure `[]string{}` etc. is zero value according to Go. Should we either make switch case behavior same as reflect's zerovalue or make zero length slices zero value?

Comment (sixcolors) [reply utils/zero.go]: https://github.com/gofiber/fiber/pull/2760/commits/c228dbd3c4c1fcc05c4508e568e31594bafa3446 aligns with conventions by removing slice cases and updating tests.

**Comment by gaby** [comment]
@sixcolors Probably worth comparing your implementation with this one: https://github.com/duke-git/lancet/blob/main/validator/validator.go#L390

Unit-Tests here: https://github.com/duke-git/lancet/blob/main/validator/validator_test.go#L413

Lancet is used by over +1k repos, so it's worth comparing.

**Comment by sixcolors** [comment]
Set to draft while considering if related change is appropriate or not

**Comment by sixcolors** [comment]
As @nickajacks1 pointed out https://github.com/gofiber/fiber/pull/2751#issuecomment-1852451213 a IsZeroValue check for ContdxtKeys would cause issues for some types that could be valid keys. So I’m closing this PR.
